File: clients/fsharp/Model/UnitOrder.fs
#nowarn "0058"

namespace AiCup22
namespace AiCup22.Model

open AiCup22

/// Order for specific unit
type UnitOrder = {
    /// Target moving velocity
    TargetVelocity: Model.Vec2;
    /// Target view direction (vector length doesn't matter)
    TargetDirection: Model.Vec2;
    /// Order to perform an action, or None
    Action: option<Model.ActionOrder>;
} with

    /// Write UnitOrder to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        this.TargetVelocity.writeTo writer
        this.TargetDirection.writeTo writer
        match this.Action with
            | Some value ->
                writer.Write true
                value.writeTo writer
            | None -> writer.Write false
        ()

    /// Read UnitOrder from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        TargetVelocity = Model.Vec2.readFrom reader;
        TargetDirection = Model.Vec2.readFrom reader;
        Action = match reader.ReadBoolean() with
                     | true -> Some(Model.ActionOrder.readFrom reader;)
                     | false -> None
    }

File: clients/csharp/Model/UnitOrder.cs
namespace AiCup22.Model
{
    /// <summary>
    /// Order for specific unit
    /// </summary>
    public struct UnitOrder
    {
        /// <summary>
        /// Target moving velocity
        /// </summary>
        public AiCup22.Model.Vec2 TargetVelocity { get; set; }
        /// <summary>
        /// Target view direction (vector length doesn't matter)
        /// </summary>
        public AiCup22.Model.Vec2 TargetDirection { get; set; }
        /// <summary>
        /// Order to perform an action, or None
        /// </summary>
        public AiCup22.Model.ActionOrder Action { get; set; }
    
        public UnitOrder(AiCup22.Model.Vec2 targetVelocity, AiCup22.Model.Vec2 targetDirection, AiCup22.Model.ActionOrder action)
        {
            this.TargetVelocity = targetVelocity;
            this.TargetDirection = targetDirection;
            this.Action = action;
        }
    
        /// <summary> Read UnitOrder from reader </summary>
        public static UnitOrder ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new UnitOrder();
            result.TargetVelocity = AiCup22.Model.Vec2.ReadFrom(reader);
            result.TargetDirection = AiCup22.Model.Vec2.ReadFrom(reader);
            if (reader.ReadBoolean())
            {
                result.Action = AiCup22.Model.ActionOrder.ReadFrom(reader);
            } else
            {
                result.Action = null;
            }
            return result;
        }
    
        /// <summary> Write UnitOrder to writer </summary>
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            TargetVelocity.WriteTo(writer);
            TargetDirection.WriteTo(writer);
            if (Action == null)
            {
                writer.Write(false);
            } else
            {
                writer.Write(true);
                Action.WriteTo(writer);
            }
        }
    
        /// <summary> Get string representation of UnitOrder </summary>
        public override string ToString() {
            string stringResult = "UnitOrder { ";
            stringResult += "TargetVelocity: ";
            stringResult += TargetVelocity.ToString();
            stringResult += ", ";
            stringResult += "TargetDirection: ";
            stringResult += TargetDirection.ToString();
            stringResult += ", ";
            stringResult += "Action: ";
            if (Action == null)
            {
                stringResult += "null";
            } else
            {
                stringResult += Action.ToString();
            }
            stringResult += " }";
            return stringResult;
        }
    }
}